grpc/conn.go

```diff
@@ -0,0 +1,67 @@
+package grpc
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"go.viam.com/utils/rpc"
+	googlegrpc "google.golang.org/grpc"
+)
+
+// ReconfigurableClientConn allows for the underlying client connections to be swapped under the hood.
+type ReconfigurableClientConn struct {
+	connMu sync.RWMutex
+	conn   rpc.ClientConn
+}
+
+// Invoke invokes using the underlying client connection.
+func (c *ReconfigurableClientConn) Invoke(
+	ctx context.Context,
+	method string,
+	args, reply interface{},
+	opts ...googlegrpc.CallOption,
+) error {
+	c.connMu.RLock()
+	conn := c.conn
+	c.connMu.RUnlock()
```

Comment:
Should we just `defer c.connMu.RUnlock()` instead (same with `NewStream`)? What if a `Close` comes in after this? Can't the `Invoke` cause a `panic`? (I realize you did not write this code and just moved it).

Reply:
Invoke would fail with `grpc: the client connection is closing` if we close the connection after making a copy. So that's not too bad. If we defer the lock, it may take forever to close, which becomes another issue. This way, we hold the lock for as short as possible

Reply:
[q] somewhat related - why do have this pattern of "copying" the connection and then using it, vs locking and using the connection field directly?

Reply:
> Invoke would fail with grpc: the client connection is closing if we close the connection after making a copy. So that's not too bad. If we defer the lock, it may take forever to close, which becomes another issue. This way, we hold the lock for as short as possible

Gotcha, I guess I was worried since we're nilling out `c.conn` in `Close`, and I thought `conn.Invoke` would fail. But, I guess, precisely because you're "copying" the connection pointer with that `conn` variable as @maximpertsov pointed out, you still have a reference to the `rpc.ClientConn`?

Reply:
I think the concern would be around closing - with making a copy the lock is held for shorter and we can close faster

Reply:
discussed briefly IRL: I wonder if it would clearer to not copy the connection and instead rely on locks directly. We would need to add a context to cancel long-running connections but maybe that pattern is a bit clearer?

again, definitely not a change for this PR